logger.ts

```diff
@@ -0,0 +1,158 @@
+import * as fs from "fs";
+import * as stream from "stream";
+import * as util from "util";
+
+import { once } from "events";
+
+enum LogDestination {
+  Console,
+  File,
+}
+
+class Logger {
+  private logDestination: LogDestination;
+  private logFileName: string;
+  private logFileStream: fs.WriteStream;
+  private defaultLogger: Logger | undefined;
+
+  // This is a promisified version of stream.finished to wait for the stream to finish writing
+  private streamFinshed = util.promisify(stream.finished);
+
+  private logPrefix = {
+    debug: "[DEBUG]:",
+    info: "[INFO]:",
+    warn: "[WARN]:",
+    error: "[ERROR]:",
+  };
+
+  constructor(
+    logDestination: LogDestination = LogDestination.Console,
+    logFileName: string = "./log.txt"
+  ) {
+    this.logDestination = logDestination;
+    this.logFileName = logFileName;
+    this.logFileStream = fs.createWriteStream(logFileName, {
+      encoding: "utf-8",
+      flags: "a",
+    });
+  }
+
+  // This allows a singleton logger to be used throughout the application if needed
+  /**
+   * call this method to get a singleton instance of the Logger class
+   * @returns {Logger} A singleton instance of the Logger class
+   */
+  getLogger() {
+    if (!this.defaultLogger) {
+      this.defaultLogger = new Logger();
+    }
+    return this.defaultLogger;
+  }
```

Comment:
is it working ? please add a test to check it. i think that in order to call this method you need an instance . so it is not really a singleton.

Reply:
?